test-support/page-object/helpers.js

```diff
@@ -1,27 +1,60 @@
 import Ember from 'ember';
+import Ceibo from './ceibo';
 
-export function qualifySelector(...selectors) {
-  return selectors.filter(item => !!item).join(' ');
+function getScopes(target) {
+  var scopes = [];
+
+  if (target.scope) {
+    scopes.push(target.scope);
+  }
+
+  if (Ceibo.parent(target)) {
+    scopes = scopes.concat(calculateScope(Ceibo.parent(target)));
+  }
+
+  return scopes;
 }
 
-export function findElementWithAssert(options, target) {
-  let selector = qualifySelector(
-    options.scope || target.scope,
-    indexedSelector(options.selector, options.index)
-  );
+function query(tree, selector, options, fn) {
+  var scope;
 
-  /* global findWithAssert */
-  return findWithAssert(selector);
+  if (options.resetScope) {
+    scope = options.scope;
+  } else {
+    scope = calculateScope(tree, options.scope);
+  }
+
+  if (!selector) {
+    selector = scope;
+    scope = undefined;
+  }
+
+  if (typeof options.at === 'number') {
+    selector = `${selector}:eq(${options.at})`;
+  }
+
+  return fn(selector, scope);
 }
 
-export function findElement(options, target) {
-  let selector = qualifySelector(
-    options.scope || target.scope,
-    indexedSelector(options.selector, options.index)
-  );
+function calculateScope(target, propertyScope) {
+  var scopes = getScopes(target);
+
+  scopes.reverse();
+  scopes.push(propertyScope);
 
-  /* global find */
-  return find(selector);
+  return $.trim(scopes.join(' '));
+}
+
+export function buildSelector(target, selector, options) {
+  return Ember.$.trim(`${calculateScope(target, options.scope)} ${selector}`);
+}
+
+export function findElementWithAssert(tree, selector, options) {
+  return query(tree, selector, options, findWithAssert);
+}
+
+export function findElement(tree, selector, options, target) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['target' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059192/issues/source?bid=2903707&fileBranchId=3076256#l56)